<html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=gbk">
<title>PostgreSQL research</title>
<style type="text/css">
.blogcnt{line-height:160%;font-size:14px;text-align:left;word-wrap:break-word;}
.blogcnt *{line-height:160%;}
.blogcnt p{margin:0 0 10px;}
.blogcnt ul,.nbw-blog ol{margin:5px 0 5px 40px;padding:0}
.blogcnt em{font-style:italic;}
.blogcnt blockquote{font-size:1em;margin:auto 0 auto 35px;}
.blogcnt img{border:0;max-width:100%;}
</style>
</head>
<body style="color:#444444;">
<h1 id="blog-Title"><a href="index.html">PostgreSQL research</a></h1>
<div id="" style="padding:0 20px;">
	<h2 id="">Lua Errors</h2>
	<h5 id="">2014-02-19 16:57:01&nbsp;&nbsp;&nbsp;<a href="http://blog.163.com/digoal@126/blog/static/16387704020141194286494/" target="_blank">查看原文&gt;&gt;</a></h5>
	<div class="" id="" style="padding:0 20px;">
		<div class="blogcnt" style="width:800px;">Lua 是一个扩展性语言, 通常嵌入在其他程序中使用.<wbr><div>所以当Lua运行错误时, 不建议简单的crash或退出, 最好是能返回错误代码和错误信息.</div><div>当lua运行时遇到某些错误会抛出异常, 例如对非数字进行数学运算, 调用不存在的函数, 使用不存在的索引取表的值. 通过metatable可以自定义这些错误.</div><div><br></div><div>除了这些常见的错误, 我们还可以在lua程序中使用error函数人为的产生错误.</div><div>例如, 当条件满足时, 使用error抛出异常 :&nbsp;</div><div><pre class="prettyprint"   ><p></p><div><font size="2"   >&gt; f = io.open("/root/test1.lua","w")</font></div><div><div><font size="2"   >&gt; f:write([[ print("enter a number:")</font></div><div><font size="2"   >&gt;&gt; n = io.read("*n")</font></div><div><font size="2"   >&gt;&gt; if not n then error("invalid number") end</font></div><div><font size="2"   >&gt;&gt; print(n)</font></div><div><font size="2"   >&gt;&gt; ]])</font></div><div><font size="2"   >&gt; f:close<span style="line-height: 28px;"   >()</span></font></div></div><div><span style="line-height: 28px;"   ><font size="2"   ><div>&gt; dofile("/root/test1.lua")</div><div>enter a number:</div><div><div>9</div><div>9</div></div><div><div>&gt; dofile("/root/test1.lua")</div><div>enter a number:</div><div>abc</div><div>/root/test1.lua:3: invalid number</div><div>stack traceback:</div><div>&nbsp; &nbsp; &nbsp; &nbsp; [C]: in function 'error'</div><div>&nbsp; &nbsp; &nbsp; &nbsp; /root/test1.lua:3: in main chunk</div><div>&nbsp; &nbsp; &nbsp; &nbsp; [C]: in function 'dofile'</div><div>&nbsp; &nbsp; &nbsp; &nbsp; stdin:1: in main chunk</div><div>&nbsp; &nbsp; &nbsp; &nbsp; [C]: in ?</div><div>后面一直报错, 不知道是不是BUG</div><div>&gt; dofile("/root/test1.lua")</div><div>enter a number:</div><div>/root/test1.lua:3: invalid number</div><div>stack traceback:</div><div>&nbsp; &nbsp; &nbsp; &nbsp; [C]: in function 'error'</div><div>&nbsp; &nbsp; &nbsp; &nbsp; /root/test1.lua:3: in main chunk</div><div>&nbsp; &nbsp; &nbsp; &nbsp; [C]: in function 'dofile'</div><div>&nbsp; &nbsp; &nbsp; &nbsp; stdin:1: in main chunk</div><div>&nbsp; &nbsp; &nbsp; &nbsp; [C]: in ?</div></div></font></span></div><p></p></pre></div><div><br></div><div>使用assert也能达到同样的效果, 评估第一个参数(<span style="line-height: 28px;"   >io.read("*n")</span><span style="line-height: 28px;"   >)的返回值, 如果是false 则抛出异常, 并且输出第二个参数作为错误消息.&nbsp;</span></div><div><span style="line-height: 28px;"   >io.read("*n") , 如果输入不是数字, 返回nil.</span></div><div><pre class="prettyprint"   ><p></p><div><font size="2"   >&gt; n = assert(io.read("*n"), "invalid number")</font></div><div><font size="2"   >99</font></div><div><font size="2"   >&gt; print(n)</font></div><div><font size="2"   >99</font></div><div><font size="2"   >&gt; n = assert(io.read("*n"), "invalid number")</font></div><div><font size="2"   >abc</font></div><div><font size="2"   >stdin:1: invalid number</font></div><div><font size="2"   >stack traceback:</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; [C]: in function 'assert'</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; stdin:1: in main chunk</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; [C]: in ?</font></div><div><font size="2"   >&gt; print(n)</font></div><div><font size="2"   >99</font></div><div><font size="2"   >&gt; n = assert(io.read("*n"), "invalid number")</font></div><div><font size="2"   >stdin:1: invalid number</font></div><div><font size="2"   >stack traceback:</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; [C]: in function 'assert'</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; stdin:1: in main chunk</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; [C]: in ?</font></div><p></p></pre></div><div><pre class="prettyprint"   ><p></p><div><font size="2"   >&gt; n = io.read("*l")</font></div><div><font size="2"   >ab he</font></div><div><font size="2"   >&gt; print(n)</font></div><div><font size="2"   >ab he</font></div><div><font size="2"   >&gt; n = io.read("*n")</font></div><div><font size="2"   >ab12c2d &nbsp; &nbsp;&nbsp;</font></div><div><font size="2"   >&gt; print(n)</font></div><div><font size="2"   >nil</font></div><div><font size="2"   >&gt; n = io.read("*n")<br>123 3<br>&gt; print(n)<br>123</font></div><p></p></pre></div><div><br></div><div>数字判断的例子还可以这么写, 使用 assert 来判断tonumber(n)是否返回nil, 返回nil报错.</div><div><pre class="prettyprint"   ><p></p><div><font size="2"   >n = io.read()</font></div><div><font size="2"   >assert(tonumber(n), "invalid input:" .. n .. " is not a number")</font></div><p></p></pre></div><div>注意assert在执行前, 先执行参数中所有的表达式, 所以不管n是否为数字, 都会执行这个字符串连接.<span style="line-height: 28px;"   >"invalid input:" .. n .. " is not a number"</span></div><div><span style="line-height: 28px;"   ><br></span></div><div>文件操作的用法 :&nbsp;</div><div>file = assert(io.open(filename, $fmt))</div><div><span style="line-height: 28px;"   >一般用r模式打开文件来</span>判断文件是否存在.</div><div><br></div><div>[参考]</div><div>1.&nbsp;</div><div><h3 style="font-family: Verdana, Geneva, sans-serif; font-weight: normal; padding-left: 0.5em; border-left-style: solid; border-left-color: rgb(208, 208, 255); border-left-width: 1em; line-height: normal; text-align: justify;"   ><a name="pdf-error" rel="nofollow"   >error (message [, level])</a></h3><span style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >Terminates the last protected function called and returns&nbsp;</span><code style="font-size: medium; line-height: normal; text-align: justify;"   >message</code><span style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >&nbsp;as the error message. Function&nbsp;</span><code style="font-size: medium; line-height: normal; text-align: justify;"   >error</code><span style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >&nbsp;never returns.</span><p style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >Usually,&nbsp;<code style="font-size: 12pt;"   >error</code>&nbsp;adds some information about the error position at the beginning of the message, if the message is a string. The&nbsp;<code style="font-size: 12pt;"   >level</code>&nbsp;argument specifies how to get the error position. With level&nbsp;1 (the default), the error position is where the&nbsp;<code style="font-size: 12pt;"   >error</code>&nbsp;function was called. Level&nbsp;2 points the error to where the function that called&nbsp;<code style="font-size: 12pt;"   >error</code>&nbsp;was called; and so on. Passing a level&nbsp;0 avoids the addition of error position information to the message.</p></div><div>2.&nbsp;</div><div><h3 style="font-family: Verdana, Geneva, sans-serif; font-weight: normal; padding-left: 0.5em; border-left-style: solid; border-left-color: rgb(208, 208, 255); border-left-width: 1em; line-height: normal; text-align: justify;"   ><a name="pdf-assert" rel="nofollow"   >assert (v [, message])</a></h3><span style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >Issues an error when the value of its argument&nbsp;</span><code style="font-size: medium; line-height: normal; text-align: justify;"   >v</code><span style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >&nbsp;is false (i.e.,&nbsp;</span><b style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >nil</b><span style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >&nbsp;or&nbsp;</span><b style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >false</b><span style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >); otherwise, returns all its arguments.&nbsp;</span><code style="font-size: medium; line-height: normal; text-align: justify;"   >message</code><span style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >&nbsp;is an error message; when absent, it defaults to "assertion failed!"</span></div><div>3.&nbsp;</div><div><h3 style="font-family: Verdana, Geneva, sans-serif; font-weight: normal; padding-left: 0.5em; border-left-style: solid; border-left-color: rgb(208, 208, 255); border-left-width: 1em; line-height: normal; text-align: justify;"   ><a name="pdf-io.read" rel="nofollow"   >io.read (・・・)</a></h3><p style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >Equivalent to&nbsp;<code style="font-size: 12pt;"   >io.input():read(・・・)</code>.</p></div><div><ul style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   ><li><b>"<code style="font-size: 12pt;"   >*n</code>":&nbsp;</b>reads a number; this is the only format that returns a number instead of a string.</li><li><b>"<code style="font-size: 12pt;"   >*a</code>":&nbsp;</b>reads the whole file, starting at the current position. On end of file, it returns the empty string.</li><li><b>"<code style="font-size: 12pt;"   >*l</code>":&nbsp;</b>reads the next line skipping the end of line, returning&nbsp;<b>nil</b>&nbsp;on end of file. This is the default format.</li><li><b>"<code style="font-size: 12pt;"   >*L</code>":&nbsp;</b>reads the next line keeping the end of line (if present), returning&nbsp;<b>nil</b>&nbsp;on end of file.</li><li><b><em>number</em>:&nbsp;</b>reads a string with up to this number of bytes, returning&nbsp;<b>nil</b>&nbsp;on end of file. If number is zero, it reads nothing and returns an empty string, or&nbsp;<b>nil</b>&nbsp;on end of file.</li></ul></div></div>
	</div>
</div>
</body>
</html>